### PR TITLE
Add wghelper to daemon implement first pieces for linux

### PR DESCRIPTION
**Pull request by birdsarah**
This is first PR that sets up the general architecture of how adding all the pieces that will allow us to remove helper scripts is going to go. This includes moving the Daemon::Config to wgutils InterfaceConfig. Although the immediate use for that isn't clear in this PR. Future methods in wgutils will need the Daemon::Config and we can end up with a circular import if we're not careful.

### CODE REVIEW

File: src/daemon/daemon.cpp [hunk 1]
```diff
@@ -82,6 +83,13 @@ bool Daemon::activate(const Config& config) {
     return activate(config);
   }
 
+  if (supportWGUtils()) {
```
Comment (bakulf) [inline src/daemon/daemon.cpp:86]: You can merge these 2 if-stmt: 
```
if (supportWGUtils() && wgutils()->interfaceExists()) { ...
```

File: src/daemon/daemon.cpp [hunk 2]
```diff
@@ -222,8 +230,15 @@ bool Daemon::deactivate(bool emitSignals) {
     return true;
   }
 
-  m_connected = false;
+  if (supportWGUtils()) {
```
Comment (bakulf) [inline src/daemon/daemon.cpp:233]: Same here: `if (a && b)`

File: src/daemon/config.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
+
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef INTERFACECONFIG_H
```
Comment (bakulf) [inline src/daemon/config.h:6]: This file should be called 'interfaceconfig.h' or here we should replace 'interfaceConfig' with 'Config'

File: src/daemon/wgutils.h [hunk 4]
```diff
@@ -0,0 +1,26 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef WIREGUARDUTILS_H
```
Comment (bakulf) [inline src/daemon/wgutils.h:5]: Here too: WireguardUtils vs WgUtils

File: src/daemon/wgutils.h [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+#define WIREGUARDUTILS_H
+
+#include <QCoreApplication>
+
+constexpr const char* WG_INTERFACE = "moz0";
+
+class WireguardUtils : public QObject {
```
Comment (bakulf) [inline src/daemon/wgutils.h:12]: Usually, the filename matches with the name of the class

File: src/platforms/linux/daemon/dbusservice.cpp [hunk 6]
```diff
@@ -40,6 +41,13 @@ DBusService::DBusService(QObject* parent) : Daemon(parent) {
 
 DBusService::~DBusService() { MVPN_COUNT_DTOR(DBusService); }
 
+WireguardUtilsLinux* DBusService::wgutils() {
+  if (!m_wgutils) {
+    m_wgutils = new WireguardUtilsLinux();
```
Comment (bakulf) [inline src/platforms/linux/daemon/dbusservice.cpp:46]: This is a bug: here you are allocating an object but you do not deallocate it.
The simplest solution is to set a parent to WireguardUtilsLinux and pass it to QObject. See this: https://doc.qt.io/qt-5/qobject.html#QObject "The destructor of a parent object destroys all child objects."
\\

Comment (bakulf) [reply src/platforms/linux/daemon/dbusservice.cpp:46]: If you add MVPN_COUNT_CTOR/DTOR in the wgutils, you will see that in debug builds the leak is reported.

File: src/platforms/linux/daemon/wgutilslinux.cpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+bool WireguardUtilsLinux::interfaceExists() {
+  // As currentInterfaces only gets wireguard interfaces, this method
+  // also confirms an interface as being a wireguard interface.
+  return currentInterfaces().contains(WG_INTERFACE);
+};
+
+QStringList WireguardUtilsLinux::currentInterfaces() {
+  char* deviceName;
+  size_t len;
+  QStringList devices;
```
Comment (bakulf) [inline src/platforms/linux/daemon/wgutilslinux.cpp:30]: Move these next to the first use. Line 35.

**Review by bakulf — CHANGES_REQUESTED** [review]

File: src/daemon/config.h (hunk 3 above)
Comment (birdsarah) [reply src/daemon/config.h:6]: What do you think of the name InterfaceConfig? It seems right, but I wanted to check. If InterfaceConfig is good, I'll rename the file.

File: src/daemon/wgutils.h (hunk 5 above)
Comment (birdsarah) [reply src/daemon/wgutils.h:12]: got it.

File: src/platforms/linux/daemon/dbusservice.cpp (hunk 6 above)
Comment (birdsarah) [reply src/platforms/linux/daemon/dbusservice.cpp:46]: Thank you. I'll fix this. Sorry I still have things to learn to get this ingrained in my head.

Comment (birdsarah) [reply src/platforms/linux/daemon/dbusservice.cpp:45]: Resolved in 9761274

**Comment by birdsarah** [comment]
@bakulf this is ready for a re-review.

I have not implemented `release` as we started talking about in slack. I believe it's not necessary in this PR as this PR doesn't do anything that needs to be un-done. It simply provides the checks and sets up the initial architecture.

File: src/daemon/interfaceconfig.h [hunk 8]
```diff
@@ -0,0 +1,25 @@
+
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (bakulf) [inline src/daemon/interfaceconfig.h:2]: No extra line here.

File: src/daemon/interfaceconfig.h [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef INTERFACECONFIG_H
+#define INTERFACECONFIG_H
+
+#include "ipaddressrange.h"
```
Comment (bakulf) [inline src/daemon/interfaceconfig.h:9]: I would include <QList> and <QString>

File: src/platforms/linux/daemon/dbusservice.cpp [hunk 10]
```diff
@@ -40,6 +40,14 @@ DBusService::DBusService(QObject* parent) : Daemon(parent) {
...
+    m_wgutils->setParent(this);
```
Comment (bakulf) [inline src/platforms/linux/daemon/dbusservice.cpp:46]: This works too, but why not passing the argument in the CTOR?

File: src/platforms/linux/daemon/dbusservice.h [hunk 11]
```diff
@@ -34,16 +35,19 @@ class DBusService final : public Daemon {
...
-  bool run(Op op, const Config& config) override;
+  bool run(Op op, const InterfaceConfig& config) override;
 
-  bool supportServerSwitching(const Config& config) const override;
+  bool supportServerSwitching(const InterfaceConfig& config) const override;
+  bool switchServer(const InterfaceConfig& config) override;
 
-  bool switchServer(const Config& config) override;
+  bool supportWGUtils() const override { return true; }
+  WireguardUtilsLinux* wgutils() override;
```
Comment (bakulf) [inline src/platforms/linux/daemon/dbusservice.h:44]: This should return a `WireguardUtils` and not a `WireguardUtilsLinux`.

File: src/platforms/linux/daemon/wireguardutilslinux.cpp [hunk 12]
```diff
@@ -0,0 +1,51 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+#include "wireguardutilslinux.h"
```
Comment (bakulf) [inline src/platforms/linux/daemon/wireguardutilslinux.cpp:4]: Empty line between the header and the first include

File: src/platforms/linux/daemon/wireguardutilslinux.cpp [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+#if defined(__cplusplus)
+}
+#endif
+// End import wireguard
+
+namespace {
+Logger logger(LOG_LINUX, "WireguardUtilsLinux");
+}
+
+WireguardUtilsLinux::WireguardUtilsLinux() : WireguardUtils() {
```
Comment (bakulf) [inline src/platforms/linux/daemon/wireguardutilslinux.cpp:22]: No need to add `: WireguardUtils()`. It's implicit.

File: src/platforms/linux/daemon/wireguardutilslinux.cpp [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+bool WireguardUtilsLinux::interfaceExists() {
+  // As currentInterfaces only gets wireguard interfaces, this method
+  // also confirms an interface as being a wireguard interface.
+  return currentInterfaces().contains(WG_INTERFACE);
+};
+
+QStringList WireguardUtilsLinux::currentInterfaces() {
+  char* deviceNames = wg_list_device_names();
+  QStringList devices;
```
Comment (bakulf) [inline src/platforms/linux/daemon/wireguardutilslinux.cpp:40]: Move this to line 44. After the if-stmt.

File: src/platforms/linux/daemon/dbusservice.cpp (hunk 10 above)
Comment (bakulf) [reply src/platforms/linux/daemon/dbusservice.cpp:46]: The issue with this approach is that we can forget to call setParent().

**Review by bakulf — CHANGES_REQUESTED** [review]
All good! Just some layout tuning. I would like to see the parenting of the WireguardUtils in the CTOR instead of via setParent(), because everywhere else in the code we use this approach.

The idea is that:
1. `WireguardUtils` needs a CTOR like this: `explicit WireguardUtils(QObject* parent) : QObject(parent) {}`
2. `WireguardUtilsLinux` will have: `explicit WireguardUtilsLinux(QObject* parent) : WireguardUtils(parent) {}`
3. Just to avoid DTOR issues, add a `virtual ~WireguardUtils() = default;` See: https://www.geeksforgeeks.org/virtual-destructor/

File: src/platforms/linux/daemon/dbusservice.h (hunk 11 above)
Comment (birdsarah) [reply src/platforms/linux/daemon/dbusservice.h:44]: Fine. I had it like this as in next PR I had added a WireguardUtilsLinux specific method

File: src/platforms/linux/daemon/wireguardutilslinux.cpp (hunk 14 above)
Comment (birdsarah) [reply src/platforms/linux/daemon/wireguardutilslinux.cpp:40]: And L42 should be QStringList() ?

File: src/daemon/interfaceconfig.h (hunk 9 above)
Comment (birdsarah) [reply src/daemon/interfaceconfig.h:9]: and?

File: src/daemon/interfaceconfig.h (hunk 8 above)
Comment (birdsarah) [reply src/daemon/interfaceconfig.h:2]: wierd

File: src/platforms/linux/daemon/dbusservice.cpp (hunk 10 above)
Comment (birdsarah) [reply src/platforms/linux/daemon/dbusservice.cpp:46]: I agree can forget to call setParent. In the other case couldn't I forget to call delete? Just parenting to QObject (via WireguardUtils) doesn't ensure object is destroyed (or maybe I got something wrong when I tried this before). Or are you thinking I make the parent Daemon in WireguardUtils.

Comment (birdsarah) [reply src/platforms/linux/daemon/dbusservice.cpp:46]: Or is the idea that when I call `new WireguardUtilsLinux`, I call it with `new WireguardUtilsLinux(this)` where this is dbusservice and that way I'm forced to call use a `this` so I won't forget.

**Review by bakulf — APPROVED** [review]

**Pull request merged**